proxy.bypass should support async functions

### Describe the bug

I'm migrating our angular projects from webpack to esbuild/vite, and ran into a problem with the vite proxy: Any non-synchronous calls in `proxy.bypass` result in errors, though they kinda work:

1. If I return a `Promise` from the bypass function, I get a compile or typing error:
```
error TS2322: Type '(req: IncomingMessage, res: ServerResponse<IncomingMessage>, proxyOptions: ProxyOptions) => false | Promise<...> | undefined' is not assignable to type '(req: IncomingMessage, res: ServerResponse<IncomingMessage>, options: ProxyOptions) => string | false | void | null | undefined'.
  Type 'false | Promise<boolean> | undefined' is not assignable to type 'string | false | void | null | undefined'.
    Type 'Promise<boolean>' is not assignable to type 'string | false | void | null | undefined'.
```

I can tell [from the code that there's no explicit support for Promises in the bypass implementation](https://github.com/vitejs/vite/blob/ef7a6a35e6827b92445e5a0c2c0022616efc80dd/packages/vite/src/node/server/middlewares/proxy.ts#L204C1-L215C10), and there are no tests, but interestingly it kinda works...

2. If I ignore / work around the type error, I get this error in the console for each proxied request:
```
8:14:57 PM [vite] http proxy error: Must provide a proper URL as target
Error: Must provide a proper URL as target
    at ProxyServer.<anonymous> (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:39678:36)
    at viteProxyMiddleware (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:40021:15)
    at call (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22418:7)
    at next (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22362:5)
    at viteCachedTransformMiddleware (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:42209:5)
    at call (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22418:7)
    at next (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22362:5)
    at cors (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22890:7)
    at file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22926:17
    at originCallback (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22916:15)
```

BUT the async call does complete successfully, and the browser receives the correct async proxied result.

3. If I add a placeholder / fake `proxy.target` value (as the error message suggests), it is used instead and resolves, and when my async bypass function completes it fails because the response has already been written. So I get 2 errors, eg:

```
8:18:45 PM [vite] http proxy error: /foo
Error: getaddrinfo ENOTFOUND not.localhost
    at GetAddrInfoReqWrap.onlookupall [as oncomplete] (node:dns:120:26)
node:_http_server:342
    throw new ERR_HTTP_HEADERS_SENT('write');
    ^

Error [ERR_HTTP_HEADERS_SENT]: Cannot write headers after they are sent to the client
    at ServerResponse.writeHead (node:_http_server:342:11)
    at Timeout._onTimeout (file:///C:/src/github/johncrim/vite-bugs/node_modules/.vite-temp/vite.config.ts.timestamp-1733717919201-ba46a9b806b2d.mjs:9:17)
    at listOnTimeout (node:internal/timers:581:17)
    at process.processTimers (node:internal/timers:519:7) { 
```

Note that webpack's proxy bypass _does_ support Promises.

### Reproduction

https://github.com/johncrim/vite-bugs

### Steps to reproduce

```
git checkout https://github.com/johncrim/vite-bugs
yarn
yarn dev --config vite.config.ts
```

Open the browser, then visit the `/foo` URL. The page displays after 3 seconds, but the error listed above is shown in the terminal.

I provided 2 forms of non-sychronous bypass functions, with and without Promises:

without Promises:
```ts
// vite.config.ts
import { defineConfig } from 'vite';

export default defineConfig({
  server: {
    proxy: {
      '/foo': {
        bypass: (req, res, options) => {
          setTimeout(() => {
            res.writeHead(200, {
              'Content-Type': 'text/plain'
            });
            res.end('Hello after 3 seconds (setTimeout)')
          }, 3000);
        },
        // The error message says a target is required, but
        // uncommenting this completely breaks the proxy.
        //target: 'http://not.localhost'
      }
    }
  }
})
```

with Promises:
```ts
// vite.config2.ts
import { defineConfig } from 'vite';

export default defineConfig({
  server: {
    proxy: {
      '/foo': {
        bypass: async (req, res, options) => {
          await timeout(3000);
          res.writeHead(200, {
            'Content-Type': 'text/plain'
          });
          res.end('Hello after 3 seconds (async timeout)')
        }
      }
    }
  }
})

function timeout(ms: number) {
  return new Promise<void>((resolve, reject) => {
    setTimeout(() => resolve(), ms);
  })
}
```

Both have the same problem. I think supporting `false | string | void | Promise<false | string | void>` for the bypass return type would be ideal, and at a minimum the `viteProxyMiddleware` fn should exit after the bypass function is called if it doesn't return a string or `false`. Ideally it would wait for the Promise to resolve, handle string and false the same way they're currently handled, and exit for any other Promise result. 

IMO `bypass` should actually bypass the proxy pipeline, except in the case of 404 or "new target URL returned".

### System Info

```shell
System:
    OS: Windows 11 10.0.22631
    CPU: (20) x64 12th Gen Intel(R) Core(TM) i9-12900H
    Memory: 6.24 GB / 31.68 GB
  Binaries:
    Node: 22.6.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.19 - C:\Program Files\nodejs\yarn.CMD
    npm: 10.8.2 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Chromium (128.0.2739.67)
    Internet Explorer: 11.0.22621.3527
  npmPackages:
    vite: ^6.0.1 => 6.0.3
```


### Used Package Manager

yarn

### Logs

I don't believe the startup info is useful - this is a vanilla starter project with 2 custom config files. The error log is:

<details>
<summary>Error info inside</summary>

```shell
  vite:proxy /foo -> undefined +0ms
8:34:18 PM [vite] http proxy error: Must provide a proper URL as target
Error: Must provide a proper URL as target
    at ProxyServer.<anonymous> (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:39678:36)
    at viteProxyMiddleware (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:40021:15)
    at call (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22418:7)
    at next (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22362:5)
    at viteCachedTransformMiddleware (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:42209:5)
    at call (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22418:7)
    at next (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22362:5)
    at cors (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22890:7)
    at file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22926:17
    at originCallback (file:///C:/src/github/johncrim/vite-bugs/node_modules/vite/dist/node/chunks/dep-yUJfKD1i.js:22916:15)
  vite:time 3003.82ms /foo +0ms
```

</details>

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vite.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite/discussions) or join our [Discord Chat Server](https://chat.vite.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

--- a/packages/vite/src/node/server/middlewares/proxy.ts
+++ b/packages/vite/src/node/server/middlewares/proxy.ts
@@ -189,6 +189,7 @@ export function proxyMiddleware(
                 }
               } catch (e) {
                 debug?.(`bypass: ${req.url} -> ${e}`)
+                return
               }
             }
 

--- a/packages/vite/src/node/server/middlewares/proxy.ts
+++ b/packages/vite/src/node/server/middlewares/proxy.ts
@@ -188,7 +188,13 @@ export function proxyMiddleware(
                   return
                 }
               } catch (e) {
-                debug?.(`bypass: ${req.url} -> ${e}`)
+                config.logger.error(
+                  `${colors.red(`ws proxy bypass error:`)}\n${err.stack}`,
+                  {
+                    timestamp: true,
+                    error: err,
+                  },
+                )
                 return
               }
             }

--- a/packages/vite/src/node/server/middlewares/proxy.ts
+++ b/packages/vite/src/node/server/middlewares/proxy.ts
@@ -33,7 +33,7 @@ export interface ProxyOptions extends HttpProxy.ServerOptions {
     | undefined
     | false
     | string
-    | Promise<void | null | undefined | false | string>
+    | Promise<void | null | undefined | boolean | string>
   /**
    * rewrite the Origin header of a WebSocket request to match the target
    *
@@ -175,16 +175,21 @@ export function proxyMiddleware(
             opts.target?.toString().startsWith('wss:')
           ) {
             if (opts.bypass) {
-              const bypassResult = await opts.bypass(req, undefined, opts)
-              if (typeof bypassResult === 'string') {
-                req.url = bypassResult
-                debug?.(`bypass: ${req.url} -> ${bypassResult}`)
+              try {
+                const bypassResult = await opts.bypass(req, undefined, opts)
+                if (typeof bypassResult === 'string') {
+                  debug?.(`bypass: ${req.url} -> ${bypassResult}`)
+                  req.url = bypassResult
+                  return
+                }
+                if (bypassResult === false) {
+                  debug?.(`bypass: ${req.url} -> 404`)
+                  socket.end('HTTP/1.1 404 Not Found\r\n\r\n', '')
+                  return
+                }
+              } catch (e) {
+                debug?.(`bypass: ${req.url} -> ${e}`)
               }
-              if (bypassResult === false) {
-                debug?.(`bypass: ${req.url} -> 404`)
-                socket.end('HTTP/1.1 404 Not Found\r\n\r\n', '')
-              }
-              return
             }
 
             if (opts.rewrite) {
@@ -208,18 +213,22 @@ export function proxyMiddleware(
         const options: HttpProxy.ServerOptions = {}
 
         if (opts.bypass) {
-          const bypassResult = await opts.bypass(req, res, opts)
-          if (typeof bypassResult === 'string') {
-            req.url = bypassResult
-            debug?.(`bypass: ${req.url} -> ${bypassResult}`)
-            next()
+          try {
+            const bypassResult = await opts.bypass(req, res, opts)
+            if (typeof bypassResult === 'string') {
+              debug?.(`bypass: ${req.url} -> ${bypassResult}`)
+              req.url = bypassResult
+              return next()
+            }
+            if (bypassResult === false) {
+              debug?.(`bypass: ${req.url} -> 404`)
+              res.statusCode = 404
+              return res.end()
+            }
+          } catch (e) {
+            debug?.(`bypass: ${req.url} -> ${e}`)
+            return next(e)
           }
-          if (bypassResult === false) {
-            debug?.(`bypass: ${req.url} -> 404`)
-            res.statusCode = 404
-            res.end()
-          }
-          return
         }
 
         debug?.(`${req.url} -> ${opts.target || opts.forward}`)

--- a/packages/vite/src/node/server/middlewares/proxy.ts
+++ b/packages/vite/src/node/server/middlewares/proxy.ts
@@ -27,7 +27,13 @@ export interface ProxyOptions extends HttpProxy.ServerOptions {
     /** undefined for WebSocket upgrade requests */
     res: http.ServerResponse | undefined,
     options: ProxyOptions,
-  ) => void | null | undefined | false | string
+  ) =>
+    | void
+    | null
+    | undefined
+    | false
+    | string
+    | Promise<void | null | undefined | false | string>
   /**
    * rewrite the Origin header of a WebSocket request to match the target
    *
@@ -158,7 +164,7 @@ export function proxyMiddleware(
   })
 
   if (httpServer) {
-    httpServer.on('upgrade', (req, socket, head) => {
+    httpServer.on('upgrade', async (req, socket, head) => {
       const url = req.url!
       for (const context in proxies) {
         if (doesProxyContextMatchUrl(context, url)) {
@@ -169,16 +175,16 @@ export function proxyMiddleware(
             opts.target?.toString().startsWith('wss:')
           ) {
             if (opts.bypass) {
-              const bypassResult = opts.bypass(req, undefined, opts)
+              const bypassResult = await opts.bypass(req, undefined, opts)
               if (typeof bypassResult === 'string') {
                 req.url = bypassResult
                 debug?.(`bypass: ${req.url} -> ${bypassResult}`)
-                return
-              } else if (bypassResult === false) {
+              }
+              if (bypassResult === false) {
                 debug?.(`bypass: ${req.url} -> 404`)
                 socket.end('HTTP/1.1 404 Not Found\r\n\r\n', '')
-                return
               }
+              return
             }
 
             if (opts.rewrite) {
@@ -194,7 +200,7 @@ export function proxyMiddleware(
   }
 
   // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
-  return function viteProxyMiddleware(req, res, next) {
+  return async function viteProxyMiddleware(req, res, next) {
     const url = req.url!
     for (const context in proxies) {
       if (doesProxyContextMatchUrl(context, url)) {
@@ -202,16 +208,18 @@ export function proxyMiddleware(
         const options: HttpProxy.ServerOptions = {}
 
         if (opts.bypass) {
-          const bypassResult = opts.bypass(req, res, opts)
+          const bypassResult = await opts.bypass(req, res, opts)
           if (typeof bypassResult === 'string') {
             req.url = bypassResult
             debug?.(`bypass: ${req.url} -> ${bypassResult}`)
-            return next()
-          } else if (bypassResult === false) {
+            next()
+          }
+          if (bypassResult === false) {
             debug?.(`bypass: ${req.url} -> 404`)
             res.statusCode = 404
-            return res.end()
+            res.end()
           }
+          return
         }
 
         debug?.(`${req.url} -> ${opts.target || opts.forward}`)

--- a/packages/vite/src/node/server/middlewares/proxy.ts
+++ b/packages/vite/src/node/server/middlewares/proxy.ts
@@ -187,7 +187,7 @@ export function proxyMiddleware(
                   socket.end('HTTP/1.1 404 Not Found\r\n\r\n', '')
                   return
                 }
-              } catch (e) {
+              } catch (err) {
                 config.logger.error(
                   `${colors.red(`ws proxy bypass error:`)}\n${err.stack}`,
                   {

--- a/playground/proxy-bypass/__tests__/proxy-bypass.spec.ts
+++ b/playground/proxy-bypass/__tests__/proxy-bypass.spec.ts
@@ -1,5 +1,5 @@
 import { expect, test, vi } from 'vitest'
-import { browserLogs, page, serverLogs } from '~utils'
+import { browserLogs, isServe, page, serverLogs } from '~utils'
 
 test('proxy-bypass', async () => {
   await vi.waitFor(() => {
@@ -12,7 +12,7 @@ test('async-proxy-bypass', async () => {
   expect(content).toContain('Hello after 4 ms (async timeout)')
 })
 
-test('async-proxy-bypass-with-error', async () => {
+test.runIf(isServe)('async-proxy-bypass-with-error', async () => {
   await vi.waitFor(() => {
     expect(serverLogs.join('\n')).toContain('bypass error')
   })

--- a/playground/proxy-bypass/__tests__/proxy-bypass.spec.ts
+++ b/playground/proxy-bypass/__tests__/proxy-bypass.spec.ts
@@ -1,8 +1,19 @@
 import { expect, test, vi } from 'vitest'
-import { browserLogs } from '~utils'
+import { browserLogs, page, serverLogs } from '~utils'
 
 test('proxy-bypass', async () => {
   await vi.waitFor(() => {
     expect(browserLogs.join('\n')).toContain('status of 404 (Not Found)')
+  })
+})
+
+test('async-proxy-bypass', async () => {
+  const content = await page.frame('async-response').content()
+  expect(content).toContain('Hello after 4 ms (async timeout)')
+})
+
+test('async-proxy-bypass-with-error', async () => {
+  await vi.waitFor(() => {
+    expect(serverLogs.join('\n')).toContain('bypass error')
   })
 })

--- a/playground/proxy-bypass/index.html
+++ b/playground/proxy-bypass/index.html
@@ -1,2 +1,4 @@
 root app<br />
 <iframe src="/nonExistentApp" style="border: 0"></iframe>
+<iframe src="/asyncResponse" name="async-response"></iframe>
+<iframe src="/asyncThrowingError"></iframe>

--- a/playground/proxy-bypass/vite.config.js
+++ b/playground/proxy-bypass/vite.config.js
@@ -1,5 +1,7 @@
 import { defineConfig } from 'vite'
 
+const timeout = (ms) => new Promise((r) => setTimeout(r, ms))
+
 export default defineConfig({
   server: {
     port: 9606,
@@ -8,6 +10,22 @@ export default defineConfig({
         target: 'http://localhost:9607',
         bypass: () => {
           return false
+        },
+      },
+      '/asyncResponse': {
+        bypass: async (_, res) => {
+          await timeout(4)
+          res.writeHead(200, {
+            'Content-Type': 'text/plain',
+          })
+          res.end('Hello after 4 ms (async timeout)')
+          return '/asyncResponse'
+        },
+      },
+      '/asyncThrowingError': {
+        bypass: async () => {
+          await timeout(4)
+          throw new Error('bypass error')
         },
       },
     },

